src/Components/Components/src/Routing/Router.cs

```diff
@@ -72,7 +73,7 @@ public class Router : IComponent, IHandleAfterRender, IDisposable
         /// <summary>
         /// Gets or sets a handler that should be called before navigating to a new page.
         /// </summary>
-        [Parameter] public EventCallback<NavigationContext> OnNavigateAsync { get; set; }
+        [Parameter] public Func<NavigationContext, Task> OnNavigateAsync { get; set; }
```

Comment:
This fixes a bug where a re-render would be triggered too soon because of the `StateHasChanged` invocation that happens after the `OnNavigateAsync` is awaited. Since we do conditional rendering depending on whether the `OnNavigateAsync` function faulted/canceled/completed successfully, we don't want it to be re-rendered before our post-processing.

Reply:
Do you need to regen the ref assembly? Also, can we bring these APIs up for review over email?

Reply:
> Do you need to regen the ref assembly?

Not anymore! Woot!

> Also, can we bring these APIs up for review over email?

Yep, will do a ad-hoc review for this.

Reply:
But doesn't this mean the host component will no longer be re-rendered? If so that seems like a bug. It's not enough just to re-render the `Router`; we need to re-render the thing that contains the router.

Reply:
Additionally this change means we lose the ability to declare the callback as a void method or one that doesn't take any parameters.